packages/projection-typeorm/src/TypeormStabilityWindowBuffer.ts

```diff
@@ -0,0 +1,204 @@
+};
+
+export interface TypeormStabilityWindowBufferProps {
+  /**
+   * 100 by default
+   */
+  compactBufferEveryNBlocks?: number;
+}
+
+export class TypeormStabilityWindowBuffer
+  implements StabilityWindowBuffer<Operators.WithNetworkInfo & WithTypeormContext>
+{
+  #tail: Cardano.Block | 'origin';
+  readonly #logger: Logger;
+  readonly #compactEvery: number;
+  readonly #tip$ = new ReplaySubject<Cardano.Block | 'origin'>(1);
+  readonly #tail$ = new ReplaySubject<Cardano.Block | 'origin'>(1);
+  readonly tip$: Observable<Cardano.Block | 'origin'>;
+  readonly tail$: Observable<Cardano.Block | 'origin'>;
+  readonly handleEvents: UnifiedProjectorOperator<
+    Operators.WithNetworkInfo & WithTypeormContext,
+    Operators.WithNetworkInfo & WithTypeormContext
+  >;
+
+  constructor({ compactBufferEveryNBlocks = 100 }: TypeormStabilityWindowBufferProps, dependencies: WithLogger) {
+    this.tip$ = this.#tip$.asObservable();
+    this.tail$ = this.#tail$.asObservable();
+    this.#compactEvery = compactBufferEveryNBlocks;
+    this.#logger = contextLogger(dependencies.logger, 'PgStabilityWindowBuffer');
+    this.handleEvents = (evt$) =>
+      evt$.pipe(
+        concatMap((evt) =>
+          from(
+            evt.eventType === ChainSyncEventType.RollForward ? this.#rollForward(evt) : this.#rollBackward(evt)
+          ).pipe(map(() => evt))
+        )
+      );
+  }
+
+  async initialize(queryRunner: QueryRunner): Promise<void> {
+    const repository = queryRunner.manager.getRepository(BlockDataEntity);
+    const [tip, tail] = await Promise.all([
+      // findOne fails without `where:`, so using find().
+      // It makes 2 queries so is not very efficient,
+      // but it should be fine for `initialize`.
+      repository.find({
+        order: { block: { slot: 'DESC' } },
+        relations: {
+          block: true
+        },
+        select: blockDataSelect,
+        take: 1
+      }),
+      repository.find({
+        order: { block: { slot: 'ASC' } },
+        relations: {
+          block: true
+        },
+        select: blockDataSelect,
+        take: 1
+      })
+    ]);
+    this.#tip$.next(tip[0]?.data || 'origin');
+    this.#setTail(tail[0]?.data || 'origin');
+  }
+
+  shutdown(): void {
+    this.#tip$.complete();
+    this.#tail$.complete();
+  }
+
+  async #rollForward(evt: RollForwardEvent<Operators.WithNetworkInfo & WithTypeormContext>) {
+    const { eventType, transactionCommit$, queryRunner, blockEntity, block } = evt;
+    const {
+      header: { blockNo }
+    } = block;
+    const repository = queryRunner.manager.getRepository(BlockDataEntity);
+    if (eventType === ChainSyncEventType.RollForward) {
+      this.#logger.debug('Add block data at height', blockNo);
+      const blockData = repository.create({
+        block: blockEntity,
+        data: block
+      });
+      await Promise.all([repository.insert(blockData), this.#deleteOldBlockData(evt)]);
+      transactionCommit$.subscribe(() => {
```

Comment:
I think `transactionCommitted$` would be a better name, as it triggers every time the transaction is committed.
`transactionCommit$` sounds like a commit was triggered, and we should take action, not that it was done.
Just a suggestion in case you see it this way too.

Reply:
Agree with @mirceahasegan. `Committed` is less vague

Reply:
2020e6ba9